### ISSUE TITLE
Infinite loop can occur due to race condition in TableView.getBestPlanItem when MULTI_THREADED enabled

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#354]
```
What steps will reproduce the problem?
1. Enable MULTI_THREADED mode
2. Create a view with some indexes on the underlying table
3. Execute many queries simultaneously on this view

What is the expected output? What do you see instead?
The use of the "indexCache" in TableView.getBestPlanItem() is not synchronized, 
and so it is possible for multiple threads to simultaneously "put" an object to 
the indexCache.  

HashMaps and LinkedHashMaps are not threadsafe, and if objects are 
simultaneously put to a map, it is possible for hashmap entries to be created 
with a HashMap$Entry having it's "next" member point to itself.  This causes an 
infinite loop to occur when indexCache.get() is called.

The simple solution is to add a "synchronized (indexCache)" block to 
TableView.getBestPlanItem ()

Please use labels and text to provide additional information.

```

Original issue reported on code.google.com by `geoff.ze...@visiercorp.com` on 19 Oct 2011 at 8:03

**Comment by GoogleCodeExporter** [comment]
```
The exact same bug also exists in ViewIndex.getCost () when it uses the cost 
cache.  This was discovered after fixing the above race condition in a local 
build of H2.
```

Original comment by `geoff.ze...@visiercorp.com` on 20 Oct 2011 at 7:37

**Comment by GoogleCodeExporter** [comment]
```
Hi,

Thanks a lot! Using the database, I couldn't reproduce the endless loop. 
However I could reproduce it when only using the SmallLRUCache. I fixed the 
problem and added a test case. I also added a facility to help verify such 
problems (org.h2.tool.SynchronizedVerifier).

Regards,
Thomas
```

Original comment by `thomas.t...@gmail.com` on 27 Oct 2011 at 5:56
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Fixed in version 1.3.161
```

Original comment by `thomas.t...@gmail.com` on 28 Oct 2011 at 3:57
- Changed state: **Fixed**
